Userland/Utilities/userdel.cpp

```diff
@@ -103,18 +141,35 @@ int main(int argc, char** argv)
         rc = 6;
     }
 
-    if (rc == 0 && chmod(temp_filename, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
+    if (rc == 0 && chmod(temp_passwd, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
         perror("chmod");
         rc = 1;
     }
 
-    if (rc == 0 && rename(temp_filename, "/etc/passwd") < 0) {
+    if (rc == 0 && chmod(temp_shadow, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
+        perror("chmod");
+        rc = 1;
+    }
+
+    if (rc == 0 && rename(temp_passwd, "/etc/passwd") < 0) {
         perror("failed to rename the temporary passwd file");
         rc = 1;
     }
 
+    if (rc == 0 && rename(temp_shadow, "/etc/shadow") < 0) {
+        perror("failed to rename the temporary shadow file");
+        rc = 1;
+    }
+
+    if (rc) {
+        if (unlink(temp_passwd) < 0) {
+            perror("unlink");
+        }
+        return rc;
+    }
+
     if (rc) {
-        if (unlink(temp_filename) < 0) {
+        if (unlink(temp_shadow) < 0) {
             perror("unlink");
         }
         return rc;
```

Comment:
`rc` doesn't get reinitialized here before or between those two conditionals.

If `rc == 0`, only the first conditional branch will ever be executed. `unlink(temp_shadow)` will not be executed due to the `return`.

```cpp
    if (rc) {
        if (unlink(temp_passwd) < 0) {
            perror("unlink");
        }
        return rc;
    }

    if (rc) {
        if (unlink(temp_shadow) < 0) {
            perror("unlink");
        }
        return rc;
    }
```

Reply:
Thanks for the catch. Should be good now.